<a href="https://colab.research.google.com/github/fralfaro/MAT281_2024/blob/main/docs/labs/lab_09.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# MAT281 - Laboratorio N°09

<a id='p1'></a>
## I.- Problema 01


<img src="https://www.svgrepo.com/show/1064/virus.svg" width = "300" align="center"/>

El **cáncer de mama** es una proliferación maligna de células epiteliales en los conductos o lobulillos mamarios. Esta enfermedad clonal surge cuando una célula adquiere, a través de mutaciones, la capacidad de dividirse sin control, formando un tumor que puede invadir tejidos vecinos y propagarse a otras partes del cuerpo.

El conjunto de datos `BC.csv` contiene información sobre pacientes con tumores (benignos o malignos) y diversas características del tumor, calculadas a partir de imágenes digitalizadas de aspirados con aguja fina (FNA) de masas mamarias. Estas características describen los núcleos celulares en la imagen y permiten diferenciar entre tumores benignos y malignos.

A continuación, cargamos el conjunto de datos:

In [1]:
# Importar librerías
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Importar herramientas de Scikit-learn
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Configuración de gráficos
%matplotlib inline
sns.set_palette("deep", desat=0.6)
sns.set(rc={'figure.figsize': (11.7, 8.27)})

# Cargar y preparar los datos
df = pd.read_csv("https://raw.githubusercontent.com/fralfaro/MAT281_2024/main/docs/labs/data/BC.csv")
df.set_index('id', inplace=True)

# Transformación de la variable objetivo
df['diagnosis'] = df['diagnosis'].replace({'M': 1, 'B': 0}).astype(int)

# Visualizar las primeras filas del DataFrame
df.head()

<ipython-input-1-99b1ef4544f8>:26: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['diagnosis'] = df['diagnosis'].replace({'M': 1, 'B': 0}).astype(int)


diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                          
842302            1        17.99         10.38          122.80     1001.0   
842517            1        20.57         17.77          132.90     1326.0   
84300903          1        19.69         21.25          130.00     1203.0   
84348301          1        11.42         20.38           77.58      386.1   
84358402          1        20.29         14.34          135.10     1297.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760          0.3001   
842517            0.08474           0.07864          0.0869   
84300903          0.10960           0.15990          0.1974   
84348301          0.14250           0.28390          0.2414   
84358402          0.10030           0.13280          0.1980   

          concave points_mean  symmetry_mean  ...  radius_worst  \
id                                            ...                 
842302                0.14710         0.2419  ...         25.38   
842517                0.07017         0.1812  ...         24.99   
84300903              0.12790         0.2069  ...         23.57   
84348301              0.10520         0.2597  ...         14.91   
84358402              0.10430         0.1809  ...         22.54   

          texture_worst  perimeter_worst  area_worst  smoothness_worst  \
id                                                                       
842302            17.33           184.60      2019.0            0.1622   
842517            23.41           158.80      1956.0            0.1238   
84300903          25.53           152.50      1709.0            0.1444   
84348301          26.50            98.87       567.7            0.2098   
84358402          16.67           152.20      1575.0            0.1374   

          compactness_worst  concavity_worst  concave points_worst  \
id                                                                   
842302               0.6656           0.7119                0.2654   
842517               0.1866           0.2416                0.1860   
84300903             0.4245           0.4504                0.2430   
84348301             0.8663           0.6869                0.2575   
84358402             0.2050           0.4000                0.1625   

          symmetry_worst  fractal_dimension_worst  
id                                                 
842302            0.4601                  0.11890  
842517            0.2750                  0.08902  
84300903          0.3613                  0.08758  
84348301          0.6638                  0.17300  
84358402          0.2364                  0.07678  

[5 rows x 31 columns]

Con base en la información presentada, responde las siguientes preguntas. Asegúrate de incluir el código necesario para realizar los análisis requeridos y proporciona una breve explicación que describa tus resultados y el proceso que seguiste.




1. Realice un análisis exploratorio del conjunto de datos.

In [2]:
# FIXME

df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
Index: 569 entries, 842302 to 92751
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   diagnosis                569 non-null    int64  
 1   radius_mean              569 non-null    float64
 2   texture_mean             569 non-null    float64
 3   perimeter_mean           569 non-null    float64
 4   area_mean                569 non-null    float64
 5   smoothness_mean          569 non-null    float64
 6   compactness_mean         569 non-null    float64
 7   concavity_mean           569 non-null    float64
 8   concave points_mean      569 non-null    float64
 9   symmetry_mean            569 non-null    float64
 10  fractal_dimension_mean   569 non-null    float64
 11  radius_se                569 non-null    float64
 12  texture_se               569 non-null    float64
 13  perimeter_se             569 non-null    float64
 14  area_se                 

diagnosis  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  569.000000   569.000000    569.000000      569.000000   569.000000   
mean     0.372583    14.127292     19.289649       91.969033   654.889104   
std      0.483918     3.524049      4.301036       24.298981   351.914129   
min      0.000000     6.981000      9.710000       43.790000   143.500000   
25%      0.000000    11.700000     16.170000       75.170000   420.300000   
50%      0.000000    13.370000     18.840000       86.240000   551.100000   
75%      1.000000    15.780000     21.800000      104.100000   782.700000   
max      1.000000    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

se observa que:


*   No hay valores nulos
*   hay 569 entradas y 31 calumnas
*   datos del tipo númerico






2. Normalizar las variables numéricas con el método **StandardScaler**.

In [30]:
# FIXME
# Variables:
X = df.drop('diagnosis', axis=1)
y = df['diagnosis']

# Normalizar con StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# transformar a dataframe
X_scaled_df = pd.DataFrame(X_scaled, columns=X.columns, index=X.index)

# Visualizar
X_scaled_df.head()

radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                               
842302       1.097064     -2.073335        1.269934   0.984375   
842517       1.829821     -0.353632        1.685955   1.908708   
84300903     1.579888      0.456187        1.566503   1.558884   
84348301    -0.768909      0.253732       -0.592687  -0.764464   
84358402     1.750297     -1.151816        1.776573   1.826229   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302           1.568466          3.283515        2.652874   
842517          -0.826962         -0.487072       -0.023846   
84300903         0.942210          1.052926        1.363478   
84348301         3.283553          3.402909        1.915897   
84358402         0.280372          0.539340        1.371011   

          concave points_mean  symmetry_mean  fractal_dimension_mean  ...  \
id                                                                    ...   
842302               2.532475       2.217515                2.255747  ...   
842517               0.548144       0.001392               -0.868652  ...   
84300903             2.037231       0.939685               -0.398008  ...   
84348301             1.451707       2.867383                4.910919  ...   
84358402             1.428493      -0.009560               -0.562450  ...   

          radius_worst  texture_worst  perimeter_worst  area_worst  \
id                                                                   
842302        1.886690      -1.359293         2.303601    2.001237   
842517        1.805927      -0.369203         1.535126    1.890489   
84300903      1.511870      -0.023974         1.347475    1.456285   
84348301     -0.281464       0.133984        -0.249939   -0.550021   
84358402      1.298575      -1.466770         1.338539    1.220724   

          smoothness_worst  compactness_worst  concavity_worst  \
id                                                               
842302            1.307686           2.616665         2.109526   
842517           -0.375612          -0.430444        -0.146749   
84300903          0.527407           1.082932         0.854974   
84348301          3.394275           3.893397         1.989588   
84358402          0.220556          -0.313395         0.613179   

          concave points_worst  symmetry_worst  fractal_dimension_worst  
id                                                                       
842302                2.296076        2.750622                 1.937015  
842517                1.087084       -0.243890                 0.281190  
84300903              1.955000        1.152255                 0.201391  
84348301              2.175786        6.046041                 4.935010  
84358402              0.729259       -0.868353                -0.397100  

[5 rows x 30 columns]

3. Realizar un método de reducción de dimensionalidad visto en clases.

In [11]:
# FIXME
#metodo de reduccion de dimensionalidad con PCA
pca = PCA(n_components=2) #reducir a 2 componentes
X_pca = pca.fit_transform(X_scaled) #Aplicar PCA a conhjunto de datos escalado obtenidos en punto 2
X_pca_df = pd.DataFrame(X_pca, columns=[f'PC{i+1}' for i in range(2)], index=X.index) #transformar a dataframe


4. Aplique al menos tres modelos de clasificación distintos. Para cada uno de los modelos escogidos, realice una optimización de los hiperparámetros. además, calcule las respectivas métricas. Concluya.

Se aplicaran los metodos de regresion logistica, arboles de decision y random forest. Para cada modelos se buscaran hiperparámetros con el metodo de GrindSearchCV.

In [25]:
# Regresion Logistica:


# datos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_pca_df, y, test_size=0.3, random_state=42, stratify=y)

# iteracion modelo de Logistic Regression
logreg = LogisticRegression(max_iter=1000)

# hiperparámetros Optimizacion
param_grid = {'C': [0.1, 1, 10, 100],  'solver': ['liblinear', 'lbfgs']}

# Buscar hiperparámetros con GridSearchCV
grid_search = GridSearchCV(logreg, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)
y_mejor = grid_search.best_estimator_
y_pred = y_mejor.predict(X_test)

# Metricas de evaluación
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Imprimir metricas de evaluación
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)








Accuracy: 0.9415204678362573
Precision: 0.9821428571428571
Recall: 0.859375
F1 Score: 0.9166666666666666


In [26]:
#ARBOLES DE DECISION

from sklearn.tree import DecisionTreeClassifier

# datos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_pca_df, y, test_size=0.3, random_state=42, stratify=y)
dtree = DecisionTreeClassifier(random_state=42)

# hiperparámetros Optimizacion
param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 5, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 5]
}

# Buscar hiperparámetros con GridSearchCV
grid_search = GridSearchCV(dtree, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)
y_mejor = grid_search.best_estimator_
y_pred = y_mejor.predict(X_test)

# Metricas de evaluación
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Imprimir Metricas de evaluación
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)

Accuracy: 0.935672514619883
Precision: 1.0
Recall: 0.828125
F1 Score: 0.905982905982906


In [29]:
#Random Forest:

# atos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_pca_df, y, test_size=0.3, random_state=42, stratify=y)

rf = RandomForestClassifier(random_state=42)

# hiperparámetros OPTIMIZACION
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [5, 10, 20],
    'criterion': ['gini', 'entropy']
}

# Realizar la búsqueda de hiperparámetros con GridSearchCV
grid_search = GridSearchCV(rf, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train, y_train)
y_mejor = grid_search.best_estimator_
y_pred = y_mejor.predict(X_test)

# Metricas de evaluación
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Imprimir metricas de evaluación

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)

Accuracy: 0.9532163742690059
Precision: 0.9827586206896551
Recall: 0.890625
F1 Score: 0.9344262295081968


Se observa que todas las metricas de evaluacion tienen valores cercanos a 1 validando la aprobacion de los modelos sin embargo el método de Random Forest tiene el Accuracy , Recall y F1 score más alto que los otros modelos por lo indica que es el modelo con mejor balance entre precision y detectar casos positivos. En conclusión se puede decir que de los 3 metodos de clasificacion aplicados el mejor es Random Forest.